src/server/system_services/master_key_manager.js

```diff
@@ -0,0 +1,246 @@
+
+    static get_instance() {
+        MasterKeysManager._instance = MasterKeysManager._instance || new MasterKeysManager();
+        return MasterKeysManager._instance;
+    }
+
+    /**
+     * 
+     * @returns {String} 
+     */
+    get_root_key_id() {
+        return ROOT_KEY;
+    }
+
+    /**
+     * 
+     * @returns {Object} 
+     */
+    get_root_key() {
+        return this.resolved_master_keys_by_id[ROOT_KEY];
+    }
+
+    async load_root_key() {
+        if (this.is_initialized) return;
+        const { NOOBAA_ROOT_SECRET } = process.env;
+        if (!NOOBAA_ROOT_SECRET) throw new Error('NON_EXISTING_ROOT_KEY');
+        const r_key = {
+            _id: 'noobaa-root-master-key',
+            cipher_key: Buffer.from(NOOBAA_ROOT_SECRET, 'base64'),
+            cipher_type: "aes-256-gcm",
+            description: "Root Master Key",
+            disabled: false
+        };
+        this.resolved_master_keys_by_id[ROOT_KEY] = r_key;
+        this.is_initialized = true;
+    }
+
+    /**
+     * 
+     * @param {Array} master_keys_by_id 
+     */
+    update_master_keys(master_keys_by_id) {
+        if (!master_keys_by_id) {
+            dbg.error('Invalid master_keys_by_id, got ', master_keys_by_id);
+            throw new Error('Invalid master_keys_by_id');
+        }
+        this.master_keys_by_id = master_keys_by_id;
+    }
+
+    /**
+     * 
+     * @param {Object} options 
+     * @returns {Object} 
+     */
+    new_master_key(options) {
+        const { description, master_key_id, cipher_type } = options;
+        const _id = db_client.new_object_id();
+        const m_key = _.omitBy({
+            _id,
+            description,
+            master_key_id: master_key_id && master_key_id !== ROOT_KEY ? db_client.parse_object_id(master_key_id) : undefined,
+            cipher_type: cipher_type || config.CHUNK_CODER_CIPHER_TYPE,
+            cipher_key: crypto.randomBytes(32),
+            cipher_iv: crypto.randomBytes(16),
+            disabled: false
+        }, _.isUndefined);
+
+        const encrypted_m_key = _.cloneDeep(m_key);
+        encrypted_m_key.cipher_key = this.encrypt_buffer_with_master_key_id(m_key.cipher_key, master_key_id);
+
+        this.resolved_master_keys_by_id[_id.toString()] = m_key;
+        this.master_keys_by_id[_id.toString()] = encrypted_m_key;
+        return encrypted_m_key;
+    }
+    /**
+     * 
+     * @param {String} _id 
+     * @returns {Object}  
+     */
+    get_master_key_by_id(_id) {
+        if (_id === ROOT_KEY) return this.get_root_key();
+        const mkey = this.master_keys_by_id[_id.toString()];
+        if (!mkey) throw new Error('NO_SUCH_KEY');
+        return this.resolved_master_keys_by_id[_id.toString()] || this._resolve_master_key(mkey);
+    }
+
+    /**
+     * 
+     * @param {Object} m_key 
+     * @returns {Object}  
+     */
+    _resolve_master_key(m_key) {
+        const m_of_mkey_id = (m_key.master_key_id && m_key.master_key_id._id) || m_key.master_key_id;
```

Comment:
You should consider handling this in a more consistent way, you shouldn't guess if objects are populated or not.

Reply:
deleted the last part of the phrase, it was in a case where the master_key of the master key id is the root secret

Reply:
just updating, the part after the || operator is necessary also for times when m_key.master_key_id._id doesn't exist which happened before loading from system store at the beginning.